### PR TITLE
expression: rewrite builtin function TIMESTAMP

### PR DESCRIPTION
**Pull request by spongedu**
for #4080

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -32,6 +32,7 @@ import (
 	"github.com/pingcap/tidb/sessionctx/variable"
 	"github.com/pingcap/tidb/terror"
 	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tidb/_vendor/src/github.com/cznic/mathutil"
```
Comment (zz-jason) [inline expression/builtin_time.go]: replace it with `github.com/cznic/mathutil`

File: expression/builtin_time.go [hunk 2]
```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
 	baseFunctionClass
 }
 
+func (c *timestampFunctionClass) getDefaultFsp4Type(tp *types.FieldType) int {
+	if tp.Tp == mysql.TypeDatetime || tp.Tp == mysql.TypeDate || tp.Tp == mysql.TypeDuration || tp.Tp == mysql.TypeTimestamp || tp.Tp == mysql.TypeNewDate {
```
Comment (zz-jason) [inline expression/builtin_time.go]: this line is too long, we can split it into two lines

File: expression/builtin_time.go [hunk 3]
```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
...
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinTimestampSig{newBaseBuiltinFunc(args, ctx)}
+	evalTps, argLen := []evalTp{tpString}, len(args)
+	if argLen == 2 {
+		evalTps = append(evalTps, tpString)
+	}
+	fsp := c.getDefaultFsp4Type(args[0].GetType())
+    if argLen == 2 {
```
Comment (zz-jason) [inline expression/builtin_time.go]: use `make check` to fix the code indent

File: expression/builtin_time.go [hunk 4]
```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
 	baseFunctionClass
 }
 
+func (c *timestampFunctionClass) getDefaultFsp4Type(tp *types.FieldType) int {
```
Comment (zz-jason) [inline expression/builtin_time.go]: s/getDefaultFsp4Type/getDefaultFsp/

File: expression/builtin_time.go [hunk 5]
```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
...
+    if argLen == 2 {
+		fsp = mathutil.Max(fsp, c.getDefaultFsp4Type(args[1].GetType()))
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpDatetime, evalTps...)
+    bf.tp.Decimal, bf.tp.Flen = fsp, 19
+	if fsp != 0 {
+		bf.tp.Flen += 1 + fsp
+	}
+
+	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]: check `err != nil` first before you use `bf`

File: expression/builtin_time.go [hunk 6]
```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
...
+// See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_timestamp
+func (b *builtinTimestamp2ArgsSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	ctx := b.ctx.GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalString(row, ctx)
+	if isNull || err != nil {
+		return types.Time{}, isNull, errors.Trace(err)
 	}
-	// If exists args[1].
-	s, err := args[1].ToString()
+	tm, err := types.ParseTime(arg0, mysql.TypeDatetime, getFsp(arg0))
```
Comment (zz-jason) [inline expression/builtin_time.go]: how about add a cast and let child 0 return time, so we can call `b.args[0].EvalTime(row, ctx)`

File: expression/builtin_time.go [hunk 7]
```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
...
+}
+
+type builtinTimestamp2ArgsSig struct {
+	baseTimeBuiltinFunc
+}
+
+// evalTime evals a builtinTimestamp2ArgsSig.
+// See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_timestamp
+func (b *builtinTimestamp2ArgsSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	ctx := b.ctx.GetSessionVars().StmtCtx
```
Comment (zz-jason) [inline expression/builtin_time.go]: to be compatible with our exists code, s/ctx/sc/

File: expression/builtin_time.go [hunk 8]
```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
...
+	arg1, isNull, err := b.args[1].EvalString(row, ctx)
+	if isNull || err != nil {
+		return types.Time{}, isNull, errors.Trace(err)
 	}
-	tmpDuration := arg0.Add(arg1)
-	result, err := tmpDuration.ConvertToTime(arg0.Type)
+	if !isDuration(arg1) {
+		return types.Time{}, true, nil
+	}
+    duration, err := types.ParseDuration(arg1, getFsp(arg1))
```
Comment (zz-jason) [inline expression/builtin_time.go]: how about add a cast and let child 1 return duration, so we can call `b.args[1].EvalDuration(row, ctx)`

File: expression/integration_test.go [hunk 9]
```diff
@@ -1098,6 +1098,17 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
 
+	result = tk.MustQuery("select timestamp('2003-12-31'), timestamp('2003-12-31 12:00:00','12:00:00');")
+	result.Check(testkit.Rows("2003-12-31 00:00:00 2004-01-01 00:00:00"))
+	result = tk.MustQuery("select timestamp(20170118123950.123), timestamp(20170118123950.999);")
+	result.Check(testkit.Rows("2017-01-18 12:39:50.123 2017-01-18 12:39:50.999"))
+	result = tk.MustQuery("select timestamp('2003-12-31', '01:01:01.01'), timestamp('2003-12-31.1234', '01:01:01.01');")
+	result.Check(testkit.Rows("2003-12-31 01:01:01.01 2003-12-31 01:01:01.1334"))
+	result = tk.MustQuery("select timestamp('2003-12-31', 1), timestamp('2003-12-31', -1);")
+	result.Check(testkit.Rows("2003-12-31 00:00:01 2003-12-30 23:59:59"))
+	result = tk.MustQuery("select timestamp('2003-12-32', '2000-12-12 01:01:01.01');")
```
Comment (zz-jason) [inline expression/integration_test.go]: add test for invalid input and zero time

File: expression/builtin_time.go (hunk 1 above)
Comment (spongedu) [reply expression/builtin_time.go]: ok

File: expression/builtin_time.go (hunk 4 above)
Comment (spongedu) [reply expression/builtin_time.go]: ok

File: expression/builtin_time.go (hunk 2 above)
Comment (spongedu) [reply expression/builtin_time.go]: ok

File: expression/builtin_time.go (hunk 3 above)
Comment (spongedu) [reply expression/builtin_time.go]: done

File: expression/builtin_time.go (hunk 5 above)
Comment (spongedu) [reply expression/builtin_time.go]: done

File: expression/builtin_time.go (hunk 7 above)
Comment (spongedu) [reply expression/builtin_time.go]: ok

File: expression/builtin_time.go (hunk 6 above)
Comment (spongedu) [reply expression/builtin_time.go]: The inserted `CAST` uses an default value 6, which is not as expected. examples are as follows:
**In mysql:**
```sql
mysql> select timestamp('2017-08-24 00:00:00.01');
Field   1:  `timestamp('2017-08-24 00:00:00.01')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     22
Max_length: 22
Decimals:   2
Flags:      BINARY 


+-------------------------------------+
| timestamp('2017-08-24 00:00:00.01') |
+-------------------------------------+
| 2017-08-24 00:00:00.01              |
+-------------------------------------+
1 row in set (0.00 sec)

mysql> select timestamp('2017-08-24 00:00:00.1234');
Field   1:  `timestamp('2017-08-24 00:00:00.1234')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     24
Max_length: 24
Decimals:   4
Flags:      BINARY 


+---------------------------------------+
| timestamp('2017-08-24 00:00:00.1234') |
+---------------------------------------+
| 2017-08-24 00:00:00.1234              |
+---------------------------------------+
1 row in set (0.00 sec)

```

**In TiDB, using a `CAST`:**
```sql
mysql> select timestamp('2017-08-24 00:00:00.01');
Field   1:  `timestamp('2017-08-24 00:00:00.01')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     22
Max_length: 22
Decimals:   2
Flags:      BINARY 


+-------------------------------------+
| timestamp('2017-08-24 00:00:00.01') |
+-------------------------------------+
| 2017-08-24 00:00:00.010000              |
+-------------------------------------+
1 row in set (0.00 sec)

mysql> select timestamp('2017-08-24 00:00:00.1234');
Field   1:  `timestamp('2017-08-24 00:00:00.1234')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     24
Max_length: 24
Decimals:   4
Flags:      BINARY 


+---------------------------------------+
| timestamp('2017-08-24 00:00:00.1234') |
+---------------------------------------+
| 2017-08-24 00:00:00.123400              |
+---------------------------------------+
1 row in set (0.00 sec)

File: expression/builtin_time.go (hunk 8 above)
Comment (spongedu) [reply expression/builtin_time.go]: ditto

File: expression/integration_test.go (hunk 9 above)
Comment (spongedu) [reply expression/integration_test.go]: done

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL.

File: expression/builtin_time.go [hunk 10]
```diff
@@ -2039,69 +2041,105 @@ type timestampFunctionClass struct {
...
+	switch cls := tp.ToClass(); cls {
+	case types.ClassInt:
+		return 0
+	case types.ClassReal, types.ClassString:
+		return types.MaxFsp
+	case types.ClassDecimal:
+		if tp.Decimal < types.MaxFsp {
+			return tp.Decimal
+		}
+		return 6
```
Comment (shenli) [inline expression/builtin_time.go]: Can we use types.MaxFsp here?

Comment (spongedu) [reply expression/builtin_time.go]: ok

**Comment by shenli** [comment]
LGTM
